floors/fetcher.go

```diff
@@ -0,0 +1,321 @@
+}
+
+func (fq *FetchQueue) Push(element interface{}) {
+	fetchInfo := element.(*FetchInfo)
+	*fq = append(*fq, fetchInfo)
+}
+
+func (fq *FetchQueue) Pop() interface{} {
+	old := *fq
+	n := len(old)
+	fetchInfo := old[n-1]
+	old[n-1] = nil
+	*fq = old[0 : n-1]
+	return fetchInfo
+}
+
+func (fq *FetchQueue) Top() *FetchInfo {
+	old := *fq
+	if len(old) == 0 {
+		return nil
+	}
+	return old[0]
+}
+
+func workerPanicHandler(p interface{}) {
+	glog.Errorf("floor fetcher worker panicked: %v", p)
+}
+
+func NewPriceFloorFetcher(config config.PriceFloors, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+	if !config.Enabled {
+		return nil
+	}
+
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(config.Fetcher.Worker, config.Fetcher.Capacity, pond.PanicHandler(workerPanicHandler)),
+		fetchQueue:      make(FetchQueue, 0, 100),
+		fetchInProgress: make(map[string]bool),
+		configReceiver:  make(chan FetchInfo, config.Fetcher.Capacity),
+		done:            make(chan struct{}),
+		cache:           freecache.NewCache(config.Fetcher.CacheSize * 1024 * 1024),
+		httpClient:      httpClient,
+		time:            &timeutil.RealTime{},
+		metricEngine:    metricEngine,
+	}
+
+	go floorFetcher.Fetcher()
+
+	return &floorFetcher
+}
+
+func (f *PriceFloorFetcher) SetWithExpiry(key string, value json.RawMessage, cacheExpiry int) {
+	f.cache.Set([]byte(key), value, cacheExpiry)
+}
+
+func (f *PriceFloorFetcher) Get(key string) (json.RawMessage, bool) {
+	data, err := f.cache.Get([]byte(key))
+	if err != nil {
+		return nil, false
+	}
+
+	return data, true
+}
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
+		return nil, openrtb_ext.FetchNone
+	}
+
+	// Check for floors JSON in cache
+	if result, found := f.Get(config.Fetcher.URL); found {
+		var fetchedFloorData openrtb_ext.PriceFloorRules
+		if err := json.Unmarshal(result, &fetchedFloorData); err != nil || fetchedFloorData.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return &fetchedFloorData, openrtb_ext.FetchSuccess
+	}
+
+	//miss: push to channel to fetch and return empty response
+	if config.Enabled && config.Fetcher.Enabled && config.Fetcher.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetcher, FetchTime: f.time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
+	floorData, fetchedMaxAge := f.fetchAndValidate(config)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", config.URL)
+		cacheExpiry := config.MaxAge
+		if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = fetchedMaxAge
```

Comment:
Nitpick: I noticed we are comparing `fetchedMaxAge` against `config.Period` twice. The first time happens inside `fetchFloorRulesFromURL`. Should we avoid this second comparison altogether by moving this logic to `fetchFloorRulesFromURL` somehow?
```
251   func (f *PriceFloorFetcher) fetchFloorRulesFromURL(config config.AccountFloorFetch) ([]byte, int, error) {
252 *-- 16 lines: ctx, cancel := context.WithTimeout(context.Background(), time.Duration(config.Timeout)*time.Millisecond)------------------------------------------
268  
269 -     var maxAge int
    +     var maxAge int = config.MaxAge
270       if maxAgeStr := httpResp.Header.Get("max-age"); maxAgeStr != "" {
271 -         maxAge, err = strconv.Atoi(maxAgeStr)
    +         headerMaxAge, err = strconv.Atoi(maxAgeStr)
272           if err != nil {
273               glog.Errorf("max-age in header is malformed for url %s", config.URL)
274 -         }
275 -         if maxAge <= config.Period {
    +         } else if headerMaxAge <= config.Period || headerMaxAge >= math.MaxInt32 {
276               glog.Errorf("Invalid max-age = %s provided, value should be valid integer and should be within (%v, %v)", maxAgeStr, config.Period, math.MaxInt32)
277 -         }
    +         } else {
                  maxAge = headerMaxAge
    +         }
278       }
279  
floors/fetcher.go
```

Reply:
Done